Sources/SPMUtility/IndexStore.swift

```diff
@@ -0,0 +1,240 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2019 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Basic
+import clibc
+
+public final class IndexStore {
+
+    public struct TestCaseClass {
+        public var name: String
+        public var module: String
+        public var methods: [String]
+    }
+
+    let api: IndexStoreAPI
+
+    var fn: indexstore_functions_t {
+        return api.fn
+    }
+
+    let store: indexstore_t
+
+    private init(store: indexstore_t, api: IndexStoreAPI) {
+        self.store = store
+        self.api = api
+    }
+
+    static public func open(store path: AbsolutePath, api: IndexStoreAPI) throws -> IndexStore {
+        if let store = try api.call({ api.fn.store_create(path.pathString, &$0) }) {
+            return IndexStore(store: store, api: api)
+        }
+        throw StringError("Unable to open store at \(path)")
+    }
+
+    public func listTests(inObjectFile object: AbsolutePath) throws -> [TestCaseClass] {
+        // Get the records of this object file.
+        let unitReader = try api.call{ fn.unit_reader_create(store, unitName(object: object), &$0) }
+        let records = try getRecords(unitReader: unitReader)
+
+        // Get the test classes.
+        let testCaseClasses = try records.flatMap{ try self.getTestCaseClasses(forRecord: $0) }
+
+        // Fill the module name and return.
+        let module = fn.unit_reader_get_module_name(unitReader).str
+        return testCaseClasses.map {
+            var c = $0
+            c.module = module
+            return c
+        }
+    }
+
+    private func getTestCaseClasses(forRecord record: String) throws -> [TestCaseClass] {
+        let recordReader = try api.call{ fn.record_reader_create(store, record, &$0) }
+
+        class TestCaseBuilder {
+            var classToMethods: [String: Set<String>] = [:]
+
+            func add(klass: String, method: String) {
+                classToMethods[klass, default: []].insert(method)
+            }
+
+            func build() -> [TestCaseClass] {
+                return classToMethods.map {
+                    TestCaseClass(name: $0.key, module: "", methods: $0.value.sorted())
+                }
+            }
+        }
+
+        let builder = Ref(TestCaseBuilder(), api: api)
+
+        let ctx = unsafeBitCast(Unmanaged.passUnretained(builder), to: UnsafeMutableRawPointer.self)
+        _ = fn.record_reader_occurrences_apply_f(recordReader, ctx) { ctx , occ -> Bool in
```

Comment:
What's the issue with using the functions that accept blocks ? \cc @benlangmuir

Reply:
There's a bug in Swift compiler that leads to a crash on linux if the block return a bool :( <rdar://problem/51629934>

Reply:
Ah, that's unfortunate.

Reply:
Is there a bugs.swift.org for this? For reference, I have a PR for compiling libIndexStore with `-fblocks` on linux, and I'm wondering if that PR should be blocked on this bug. In our testing on linux, I don't believe we saw any crashes. https://github.com/apple/swift-clang/pull/291